File: app/src/main/java/com/raju/javabaseproject/ui/adapters/UserAdapter.java
package com.raju.javabaseproject.ui.adapters;

import com.raju.javabaseproject.ui.adapters.delegate.UserDelegate;
import com.raju.javabaseproject.ui.adapters.delegate.base.AdapterDelegate;
import com.raju.javabaseproject.ui.adapters.delegate.base.DelegatingListAdapter;
import com.raju.javabaseproject.ui.adapters.delegate.base.ListAdapterDelegate;
import com.raju.javabaseproject.utlities.image.ImageRequester;

import javax.inject.Inject;

public class UserAdapter extends DelegatingListAdapter {

    private ImageRequester imageRequester;

    @Inject
    UserAdapter(ImageRequester imageRequester) {
        this.imageRequester = imageRequester;
    }

    @Override
    protected AdapterDelegate[] createDelegates() {
        return new ListAdapterDelegate[]{
                clickable(new UserDelegate(imageRequester))
        };
    }
}

File: app/src/main/java/com/raju/javabaseproject/ui/adapters/UserAdapter.kt
package com.raju.javabaseproject.ui.adapters

import com.raju.javabaseproject.data.model.ListItem
import com.raju.javabaseproject.ui.adapters.delegate.UserDelegate
import com.raju.javabaseproject.ui.adapters.delegate.base.AdapterDelegate
import com.raju.javabaseproject.ui.adapters.delegate.base.DelegatingListAdapter
import com.raju.javabaseproject.ui.adapters.delegate.base.ListAdapterDelegate
import com.raju.javabaseproject.utlities.image.ImageRequester

import javax.inject.Inject

class UserAdapter @Inject
internal constructor() : DelegatingListAdapter<ListItem>() {

    @Inject
    lateinit var imageRequester: ImageRequester

    @Suppress("UNCHECKED_CAST")
    override fun createDelegates(): Array<AdapterDelegate<MutableList<ListItem>>> {
        val delegate1: ListAdapterDelegate<ListItem> = clickable(UserDelegate(imageRequester) as ListAdapterDelegate<ListItem>)
        return arrayOf(delegate1)
    }
}